checklist/src/main/scala/checklist/Checked.scala

```diff
@@ -0,0 +1,138 @@
+          _ => Errored(errs, warns)
+        )
+      case Warned(warns, result) =>
+        other.fold(
+          (otherErrs, otherWarns) => Errored(otherErrs, warns.toList ++ otherWarns),
+          (otherWarns, otherResult) => Warned(warns |+| otherWarns, f(result, otherResult)),
+          otherResult => Warned(warns, f(result, otherResult))
+        )
+      case Succeeded(result) =>
+        other.fold(
+          Errored.apply _,
+          (warns, otherResult) => Warned(warns, f(result, otherResult)),
+          otherResult => Succeeded(f(result, otherResult))
+        )
+    }
+
+  def map[S](f: R => S): Checked[E, W, S] =
+    fold(
+      Errored.apply _,
+      (w, r) => Warned(w, f(r)),
+      r => Succeeded(f(r))
+    )
+
+  def mapWarnings[X](f: W => X): Checked[E, X, R] =
+    fold(
+      (e, ws) => Errored(e, ws.map(f)),
+      (ws, r) => Warned(ws.map(f), r),
+      Succeeded(_)
+    )
+
+  def mapErrors[F](f: E => F): Checked[F, W, R] =
+    fold(
+      (es, w) => Errored(es.map(f), w),
+      Warned(_, _),
+      Succeeded(_)
+    )
+
+  def flatMap[S](f: R => Checked[E, W, S]): Checked[E, W, S] =
+    fold(
+      Errored(_, _),
+      (_, result) => combineWith(f(result))((_, s) => s),
+      result => combineWith(f(result))((_, s) => s)
+    )
+
+  def value: Option[R] = fold((_, _) => None, (_, r) => Some(r), Some(_))
+  def warnings: List[W] = fold((_, w) => w, (w, _) => w.toList, _ => Nil)
+  def errors: List[E] = fold((errs, _) => errs.toList, (_, _) => Nil, _ => Nil)
+
+  def iorValidated: List[W] Ior Validated[NonEmptyList[E], R] =
+    fold(
+      (es, ws) => Ior.both(ws, Validated.invalid(es)),
+      (ws, result) => Ior.both(ws.toList, Validated.valid(result)),
+      result => Ior.right(Validated.valid(result))
+    )
+
+  def destructured: (List[E], List[W], Option[R]) =
+    fold(
+      (errs, warns) => (errs.toList, warns, None),
+      (warns, result) => (Nil, warns.toList, Some(result)),
+      result => (Nil, Nil, Some(result))
+    )
+
+  def isValid: Boolean = fold((_, _) => false, (_, _) => false, _ => true)
+  def hasAValue: Boolean = fold((_, _) => false, (_, _) => true, _ => true)
+  def hasErrors: Boolean = fold((_, _) => true, (_, _) => false, _ => false)
+  def hasNoErrors: Boolean = fold((_, _) => false, (_, _) => true, _ => true)
+  def hasWarnings: Boolean = fold((_, w) => w.nonEmpty, (_, _) => true, _ => false)
+  def hasNoWarnings: Boolean = fold((_, w) => w.isEmpty, (_, _) => false, _ => true)
+}
+
+object Checked {
+  def succeeded[E, W, R](result: R): Checked[E, W, R] = Succeeded(result)
+
+  def errored[E, W, R](errors: NonEmptyList[E]): Checked[E, W, R] = Errored(errors, Nil)
+  def erroredWithWarnings[E, W, R](errors: NonEmptyList[E], warnings: List[W]): Checked[E, W, R] = Errored(errors, warnings)
+
+  def warned[E, W, R](warnings: NonEmptyList[W], result: R): Checked[E, W, R] = Warned(warnings, result)
+
+  implicit def checkedMonad[E, W]: Monad[Checked[E, W, ?]] =
```

Comment:
I'm afraid we might be violating some laws here. Will see when I add those tests.

Reply:
Undoubtedly. IMO there are two approaches:

1. Violate those laws... who needs them anyway, right?
2. Make the implementation law abiding and provide an instance of the `Parallel` type class to implement error accumulating semantics.

I only found out about `Parallel` recently. Apologies if you know this stuff already...

`Parallel` provides a bidirectional transform between a monadic type and a non-monadic equivalent. For example, there's an instance to map back and forth between `Either` and `Validated`. Cats provides a few parallel-speicific methods that convert to the non-monadic type, do some operation, and convert back. For example:

```scala
import cats.implicits._

val either1: Either[List[String, Int] = ???
val either2: Either[List[String, Int] = ???

// Standard monadic behaviour -- fail cast:
(either1, either2).mapN(_ + _)

// Parallel behaviour -- error accumulating:
(either1, either2).parMapN(_ + _)
```

IIRC the key bits of syntax are `parMapN`, `parSequence`, and `parTraverse`.

Reply:
Definitely option (2). That way you leave the choice up to the programmer (although presumably make it very clear in the docs that `parMapN` is probably what you're looking for, since you want as many validations to run as possible).